src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,100 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size = 0;
+    private transient T[] elementData;
+
+    public ArrayList() {
+        this.elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
 
     @Override
     public void add(T value) {
-
+        elementData = expandArray();
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is not within the array");
+        }
+        elementData = expandArray();
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexExist(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        isIndexExist(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexExist(index);
+        final T value = elementData[index];
+        T[] newElementData = elementData;
+        elementData = (T[]) new Object[newElementData.length - 1];
+        System.arraycopy(newElementData, 0, elementData, 0, index);
```

Comment:
we don't need to make the capacity smaller when we remove the element

let's just shift the elements within the same array

Reply:
done ✔️ I had some trouble with this method.